src/deepqmc/physics.py

```diff
@@ -31,21 +32,208 @@ def pairwise_self_distance(coords, full=False):
+    Z_eff = mol.charges - mol.ns_core  # effective charge of the nuclei
+    effective_coulomb_potential = -(Z_eff / dists).sum(axis=(-1, -2))
+    if not mol.any_pp:
+        return effective_coulomb_potential
+
+    loc_params = mol.pp_loc_params
+    idxs = mol.pp_mask  # indices of atoms for whom we use pseudopotential
+
+    r_en = dists[:, idxs]  # electron-nucleus distances for all electrons and PP nuclei
+
+    coulomb_term = jnp.einsum(
+        'ij,ki->kji', loc_params[idxs, 0, 1, :], 1 / r_en
+    ) * jnp.exp(jnp.einsum('ij,ki->kji', -loc_params[idxs, 0, 0, :], r_en**2))
+    const_term = jnp.einsum(
+        'ij,kji->kji',
+        loc_params[idxs, 1, 1, :],
+        jnp.exp(jnp.einsum('ij,ki->kji', -loc_params[idxs, 1, 0, :], r_en**2)),
+    )
+    linear_term = jnp.einsum('ij,ki->kji', loc_params[idxs, 2, 1, :], r_en) * jnp.exp(
+        jnp.einsum('ij,ki->kji', -loc_params[idxs, 2, 0, :], r_en**2)
+    )
+
+    # Summation is carried over:
+    # - individual cores in the molecule (idxs dimension)
+    # - individual electrons (1st dimension of 'dists')
+    # - potentially over different coeffs for the term of the same type (ccECP case)
+    pseudopotential = (coulomb_term + const_term + linear_term).sum(axis=(-1, -2, -3))
+
+    return effective_coulomb_potential + pseudopotential
+
+
+@jax.jit
+def get_unit_icosahedron_sph():
+    # Basic definition of unit icosahedron vertices in spherical coordinates.
+    unit_icosahedron_sph = []
+    unit_icosahedron_sph.append([0, 0])
+    unit_icosahedron_sph.append([jnp.pi, 0])
+    for j in range(5):
+        unit_icosahedron_sph.append([jnp.arctan(2), jnp.pi / 5 * 2 * j])
+        unit_icosahedron_sph.append([jnp.pi - jnp.arctan(2), jnp.pi / 5 * (2 * j - 1)])
+    return jnp.array(unit_icosahedron_sph)
+
+
+# UNIT_ICOSAHEDRON is (12,3) array of unit icosahedron vertices
+UNIT_ICOSAHEDRON = sph2cart(get_unit_icosahedron_sph())
+QUADRATURE_THETAS = get_unit_icosahedron_sph()[:, 0]
+
+
+def get_quadrature_points(nucleus_position, rs):
+    """
+    Transform electron configuration 'rs' of size (N,3) into an array quadrature points.
+
+    Return array of size (N,12,N,3) that includes all the 12 quadrature point
+    configurations (i.e. reference electron position is shifted to another icosahedron
+    vertex) corresponding to N different reference electron.
+    """
+
+    norm = jnp.linalg.norm(rs - nucleus_position, axis=-1)
+    theta = jnp.arccos((rs - nucleus_position)[..., 2] / norm)
+    phi = jnp.arctan2((rs - nucleus_position)[..., 1], (rs - nucleus_position)[..., 0])
+
+    y_rot = jnp.moveaxis(rot_y(theta), -1, -3)  # shape: (3,3,num_e) -> (num_e,3,3)
+
+    z_rot = jnp.moveaxis(rot_z(phi), -1, -3)  # shape: (3,3,num_e) -> (num_e,3,3)
+
+    # auxiliary function applying the rotation and translation to be vmapped
+    def transform_coordinates(norm, z_rot, y_rot, r, nucleus_position):
+        return norm * (z_rot @ y_rot @ r) + nucleus_position
```

Comment:
This way of rotating the coordinates gives a fixes orientation of the quadrature points around the z axis of the icosahedron, based on its initialization. In principle there is some ambiguity, becaus a random rotation around the z-axis could be added befor the rotation around y. For a spherically symmetric system (e.g. an atom) fixing this ambiguity should not have a negative effect. However, it would be good to tests wether this can introduce a bias with respect to the molecular coordinates if a non spherically symmetric system is considerd (e.g. a diatomic molecule). If some  dependency on the relative initalization of the icosahedron wrt. the molecular coordinates can be observed I would suggest to add in a random rotation around z (like `(z_rot @ y_rot @ z_rot_random @ r)`).

Reply:
This is a very good catch. I didn't test how significant the bias introduced by the preferred orientation is. But I believe that the bias could be present even for spherically symmetric systems, since the wavefunction need not have this symmetry. In 4dbef2bd8d0762aa563c1b28b2f04be6ff1727a2 I added the random rotation about the z-axis to be on the safe side. 

Note: I also had to add some non-trivial manipulations with PRNGkeys in the `fit_wf() `function to get a randomized rotation

Reply:
Thanks! I guess you are right and in principle there could be a bias even if the Hamiltonian has some symmetry. The commit adding the random rotation looks good to me.